posts/2024-05-22-multi-dim-arrays.qmd

```diff
@@ -0,0 +1,283 @@
+---
+title: "Multi-dimensional array indexing"
+description: "Recommendations to ensure that array data and
+    index/coordinate data can be consistently interpreted
+    across software using different indexing conventions."
+author:
+    - name: "John Bogovic"
+    - name: "Davis Bennett"
+    - name: "Michael Innerberger"
+    - name: "Stephan Saalfeld"
+    - name: "Virginia Scarlett"
+date: "5/10/2024"
+---
+
+## Summary and recommendations
+
+Conventions for indexing and storing multi-dimensional data vary across programming languages and software libraries. 
+Data sharing across software with different systems is technically simple *so long as the convention used is clearly
+documented.*  For example, one software library might store and save point coordinates as `(x,y,z)`, while another
+stores them as `(z,y,x)`. Interoperability is easy (just "reverse the numbers") only if it is possible to determine
+what convention was used to produce the data, and software can determine what convention is used only if it is
+clearly and explicitly communicated.
+
+To enable seamless data and metadata sharing, software and storage formats should clearly and explicitly communicate:
+
+1. the relationship of array indexes to memory layout
+2. the semantic meaning of array dimensions
+3. the relationship of coordinate data to array dimensions
+
+Ambiguities should be clarified by (1) explicitly documenting the array indexing convention (most often C-order or F-order),
+and (2) referring to array dimensions and (3) coordinates with a consistent order and/or with labels that are
+independent of order (e.g. `['x', 'y', 'z', 't', ...]`, see
+[xarray](https://docs.xarray.dev/en/latest/getting-started-guide/why-xarray.html))
+
+::: {.callout-tip collapse="true" appearance="minimal"}
+# Examples
+
+These are provided as good examples, not as the only way to communicate the above information.
+
+#### Array size and dimension interpretation
+
+> 3D displacement fields are stored as 4D arrays, whose dimensions are ordered `[3,X,Y,Z]` where the first dimension contains
+> displacement vector components, and varies fastest in memory (i.e. F-order).
+
+#### Point coordinates
+
+> Points are stored as 3D coordinates in a CSV file ordered `X,Y,Z`
+> where the `X` dimension varies fastest in memory.
+
+:::
+
+
+## Multi-dimensional array indexing and memory layout
+
+Elements of an n-dimensional array are indexed by an ordered n-tuple, each element of which indexes into one of the array's
+dimensions. We will also call elements of this tuple "indexes." For example, the tuple `(i,j,k)` indexes a three-dimensional
+array where `i` is the "first" index, and `k` is the "last" index. Here, we will consider only the non-negative integers as
+valid indexes for arrays, though different contexts may use a different index set.
+
+Multi-dimensional arrays are often stored as one-dimensional (1D), or "flat," arrays that are interpreted, or "reshaped," into
```

Comment:
I think it's actually obligatory to store arrays in a 1D representation

Reply:
By flat / 1D I mean that they're stored in contiguous memory, and that's not necessarily true of zarr arrays for example.  While every chunk may be contiguous 1D, the "whole array" need not be.